# Подготовка файлов и программ

Некоторые программы ставятся дважды из разных источников для демонстрации различных вариантов в случае неполадок. Для оптимизации времени работы необходимо удалить лишний код.

In [1]:
pip install -q condacolab

In [4]:
import condacolab

In [5]:
condacolab.install()

✨🍰✨ Everything looks OK!


In [6]:
!conda install -c bioconda seqtk

Solving environment: / - \ | / - \ done

## Package Plan ##

  environment location: /usr/local

  added / updated specs:
    - seqtk


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    ca-certificates-2021.10.8  |       ha878542_0         139 KB  conda-forge
    certifi-2021.10.8          |   py37h89c1867_1         145 KB  conda-forge
    conda-4.11.0               |   py37h89c1867_1         983 KB  conda-forge
    seqtk-1.3                  |       h5bf99c6_3          38 KB  bioconda
    ------------------------------------------------------------
                                           Total:         1.3 MB

The following NEW packages will be INSTALLED:

  seqtk              bioconda/linux-64::seqtk-1.3-h5bf99c6_3

The following packages will be UPDATED:

  ca-certificates                      2020.12.5-ha878542_0 --> 2021.10.8-ha878542_0
  certifi                      

In [7]:
!conda install -c bioconda trimmomatic

Solving environment: - \ | / - \ | / - \ | done

## Package Plan ##

  environment location: /usr/local

  added / updated specs:
    - trimmomatic


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    openjdk-8.0.312            |       h7f98852_0        97.6 MB  conda-forge
    python_abi-3.7             |          2_cp37m           4 KB  conda-forge
    trimmomatic-0.39           |       hdfd78af_2         144 KB  bioconda
    ------------------------------------------------------------
                                           Total:        97.8 MB

The following NEW packages will be INSTALLED:

  openjdk            conda-forge/linux-64::openjdk-8.0.312-h7f98852_0
  trimmomatic        bioconda/noarch::trimmomatic-0.39-hdfd78af_2

The following packages will be UPDATED:

  python_abi                                    3.7-1_cp37m --> 3.7-2_cp37m



trimmomatic-0.39  

In [8]:
!pip install macs2 intervene

     |████████████████████████████████| 2.1 MB 3.2 MB/s 
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
     |████████████████████████████████| 5.5 MB 23.5 MB/s 
     |████████████████████████████████| 12.5 MB 31.3 MB/s 
     |████████████████████████████████| 11.2 MB 51.7 MB/s 
     |████████████████████████████████| 11.3 MB 52.5 MB/s 
  Using cached numpy-1.21.5-cp37-cp37m-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (15.7 MB)
     |████████████████████████████████| 38.1 MB 1.2 MB/s 
     |████████████████████████████████| 292 kB 52.6 MB/s 
     |████████████████████████████████| 40 kB 4.8 MB/s 
     |████████████████████████████████| 4.3 MB 45.1 MB/s 
     |████████████████████████████████| 247 kB 60.1 MB/s 
     |████████████████████████████████| 98 kB 6.2 MB/s 
     |████████████████████████████████| 1.1 MB 43.9 MB/s 
     |████████████████████████████████| 895 kB 46.1 MB/s 
     |████████████████████

In [9]:
!apt-get install bedtools

Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following NEW packages will be installed:
  bedtools
0 upgraded, 1 newly installed, 0 to remove and 39 not upgraded.
Need to get 577 kB of archives.
After this operation, 2,040 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 bedtools amd64 2.26.0+dfsg-5 [577 kB]
Fetched 577 kB in 1s (786 kB/s)
Selecting previously unselected package bedtools.
(Reading database ... 155332 files and directories currently installed.)
Preparing to unpack .../bedtools_2.26.0+dfsg-5_amd64.deb ...
Unpacking bedtools (2.26.0+dfsg-5) ...
Setting up bedtools (2.26.0+dfsg-5) ...


In [10]:
!apt-get install bowtie2 samtools

Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  cwltool libc-ares2 libhts2 libjs-bootstrap nodejs nodejs-doc
  python-asn1crypto python-avro python-cachecontrol python-certifi
  python-cffi-backend python-chardet python-cryptography python-enum34
  python-html5lib python-idna python-ipaddress python-isodate python-lockfile
  python-mistune python-openssl python-pkg-resources python-pyparsing
  python-rdflib python-rdflib-jsonld python-requests python-ruamel.yaml
  python-schema-salad python-shellescape python-six python-sparqlwrapper
  python-typing python-urllib3 python-webencodings
Suggested packages:
  bowtie2-examples python-cryptography-doc python-cryptography-vectors
  python-enum34-doc python-genshi python-lxml python-lockfile-doc
  python-openssl-doc python-openssl-dbg python-setuptools python-pyparsing-doc
  python-rdflib-doc python-rdflib-tools python-socks python-ntlm
The f

In [11]:
!conda install -c bioconda bowtie2

Solving environment: \ | / - \ | / - \ | / done

## Package Plan ##

  environment location: /usr/local

  added / updated specs:
    - bowtie2


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    bowtie2-2.3.5.1            |   py37he513fc3_0        12.0 MB  bioconda
    perl-5.32.1                | 0_h7f98852_perl5        14.5 MB  conda-forge
    tbb-2020.2                 |       h4bd325d_4         1.5 MB  conda-forge
    ------------------------------------------------------------
                                           Total:        27.9 MB

The following NEW packages will be INSTALLED:

  bowtie2            bioconda/linux-64::bowtie2-2.3.5.1-py37he513fc3_0
  perl               conda-forge/linux-64::perl-5.32.1-0_h7f98852_perl5
  tbb                conda-forge/linux-64::tbb-2020.2-h4bd325d_4



bowtie2-2.3.5.1      | 12.0 MB   | : 100% 1.0/1 [00:02<00:00,  2.30s

## Установка FastQC

In [12]:
!java -version

openjdk version "1.8.0_312"
OpenJDK Runtime Environment (Zulu 8.58.0.13-CA-linux64) (build 1.8.0_312-b07)
OpenJDK 64-Bit Server VM (Zulu 8.58.0.13-CA-linux64) (build 25.312-b07, mixed mode)


In [13]:
!wget https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.9.zip

--2022-03-10 06:46:38--  https://www.bioinformatics.babraham.ac.uk/projects/fastqc/fastqc_v0.11.9.zip
Resolving www.bioinformatics.babraham.ac.uk (www.bioinformatics.babraham.ac.uk)... 149.155.133.4
Connecting to www.bioinformatics.babraham.ac.uk (www.bioinformatics.babraham.ac.uk)|149.155.133.4|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 10249221 (9.8M) [application/zip]
Saving to: ‘fastqc_v0.11.9.zip’

fastqc_v0.11.9.zip  100%[===================>]   9.77M  2.70MB/s    in 3.6s    

2022-03-10 06:46:42 (2.70 MB/s) - ‘fastqc_v0.11.9.zip’ saved [10249221/10249221]



In [14]:
!unzip fastqc_v0.11.9.zip

Archive:  fastqc_v0.11.9.zip
  inflating: FastQC/cisd-jhdf5.jar   
   creating: FastQC/Configuration/
  inflating: FastQC/Configuration/adapter_list.txt  
  inflating: FastQC/Configuration/contaminant_list.txt  
  inflating: FastQC/Configuration/limits.txt  
  inflating: FastQC/fastqc           
  inflating: FastQC/fastqc_icon.ico  
   creating: FastQC/Help/
   creating: FastQC/Help/1 Introduction/
   creating: FastQC/Help/1 Introduction/.svn/
  inflating: FastQC/Help/1 Introduction/.svn/entries  
   creating: FastQC/Help/1 Introduction/.svn/props/
   creating: FastQC/Help/1 Introduction/.svn/text-base/
  inflating: FastQC/Help/1 Introduction/.svn/text-base/1.1 What is FastQC.html.svn-base  
   creating: FastQC/Help/1 Introduction/.svn/tmp/
   creating: FastQC/Help/1 Introduction/.svn/tmp/props/
  inflating: FastQC/Help/1 Introduction/1.1 What is FastQC.html  
   creating: FastQC/Help/2 Basic Operations/
   creating: FastQC/Help/2 Basic Operations/.svn/
  inflating: FastQC/Help/2 Basic

In [15]:
!chmod a+x FastQC/fastqc

In [16]:
!./FastQC/fastqc --help


            FastQC - A high throughput sequence QC analysis tool

SYNOPSIS

	fastqc seqfile1 seqfile2 .. seqfileN

    fastqc [-o output dir] [--(no)extract] [-f fastq|bam|sam] 
           [-c contaminant file] seqfile1 .. seqfileN

DESCRIPTION

    FastQC reads a set of sequence files and produces from each one a quality
    control report consisting of a number of different modules, each one of 
    which will help to identify a different potential type of problem in your
    data.
    
    If no files to process are specified on the command line then the program
    will start as an interactive graphical application.  If files are provided
    on the command line then the program will run with no user interaction
    required.  In this mode it is suitable for inclusion into a standardised
    analysis pipeline.
    
    The options for the program as as follows:
    
    -h --help       Print this help file and exit
    
    -v --version    Print the version of the program and exit

# Выравнивание

## Скачиваем чтения

Для примера чтения были взяты из эксперимента https://www.encodeproject.org/experiments/ENCSR000AKF/

In [ ]:
# ChIP-seq на гистоновой метке
!wget https://www.encodeproject.org/files/ENCFF000ASM/@@download/ENCFF000ASM.fastq.gz
!wget https://www.encodeproject.org/files/ENCFF000ATK/@@download/ENCFF000ATK.fastq.gz

--2022-03-01 19:52:03--  https://www.encodeproject.org/files/ENCFF000ASM/@@download/ENCFF000ASM.fastq.gz
Resolving www.encodeproject.org (www.encodeproject.org)... 34.211.244.144
Connecting to www.encodeproject.org (www.encodeproject.org)|34.211.244.144|:443... connected.
HTTP request sent, awaiting response... 307 Temporary Redirect
Location: https://encode-public.s3.amazonaws.com/2010/11/16/9e6a6ac5-1dd3-4809-850e-1b0734d54a05/ENCFF000ASM.fastq.gz?response-content-disposition=attachment%3B%20filename%3DENCFF000ASM.fastq.gz&AWSAccessKeyId=ASIATGZNGCNXX4VPAH6H&Signature=CdCTDg1f9AoL%2B0rOGe%2BtYXvBW1I%3D&x-amz-security-token=IQoJb3JpZ2luX2VjELT%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJHMEUCIQCpVh%2B7Yb5ElaLWoDpKsoUXbAmT%2BbPqh89S7oEIrjaYjgIgc3ekyuPQkUKBKo0ub4tI6BvC6oLPQ%2BCQsLwVYIcY4P8q%2BgMIHBAAGgwyMjA3NDg3MTQ4NjMiDM7mlGowGi3H%2FFfj3SrXA4rtUMFNS6qI8jBJanpHwUUSFgBbajnd7houZ56PIKJTIdCd6%2BD7px%2FKGcabyQosbv6HxdYzO9RWI2Nqbe4lwo8dsS8SvS6s1Mle97I1p99RZz2NY9J5tmpGzKv4VdbqeHFLPkYchZgAO

In [17]:
# ChIP-seq контроль
!wget https://www.encodeproject.org/files/ENCFF000ARK/@@download/ENCFF000ARK.fastq.gz

--2022-03-10 06:47:05--  https://www.encodeproject.org/files/ENCFF000ARK/@@download/ENCFF000ARK.fastq.gz
Resolving www.encodeproject.org (www.encodeproject.org)... 34.211.244.144
Connecting to www.encodeproject.org (www.encodeproject.org)|34.211.244.144|:443... connected.
HTTP request sent, awaiting response... 307 Temporary Redirect
Location: https://encode-public.s3.amazonaws.com/2010/11/16/55a84578-56e7-4ea7-a60a-7a150e21201c/ENCFF000ARK.fastq.gz?response-content-disposition=attachment%3B%20filename%3DENCFF000ARK.fastq.gz&AWSAccessKeyId=ASIATGZNGCNXSWXXRSXN&Signature=GztOQtN9z4SsFpJyER2nthYZhlw%3D&x-amz-security-token=IQoJb3JpZ2luX2VjEH8aCXVzLXdlc3QtMiJHMEUCIFHtTaCUqSYv5hUxZm9gBAyNVRd5tBC0pXOArnIiuMgjAiEAon5J%2BiF1kuySaiX5aflju%2BbxgvU4wP9O9Q8aAS1MNqMqgwQI6P%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FARAAGgwyMjA3NDg3MTQ4NjMiDLC8bI3F9NQLfmPbHirXAwObXb%2BXBI0MFBcjG47WteIPLGkUcjwGPXWhpvZZnLB%2FlVCC%2Fbkbj8mUDBbiL8v%2Fdl6ZJeXMk3SfR9%2FbfFlQx0iFtHoOU5rNNI%2BgO8irCg4ebNgzvC7bQ28PlbZVoMsLXX9KEv6%2BhKm7z

In [ ]:
# распаковка ахрива
!gzip -d /content/ENCFF000ASM.fastq.gz /content/ENCFF000ATK.fastq.gz /content/ENCFF000ARK.fastq.gz

## FastQC (или multiQC может быть использован в работе)

In [ ]:
!./FastQC/fastqc /content/ENCFF000ASM.fastq
!./FastQC/fastqc /content/ENCFF000ATK.fastq
!./FastQC/fastqc /content/ENCFF000ARK.fastq

Started analysis of ENCFF000ASM.fastq
Approx 5% complete for ENCFF000ASM.fastq
Approx 10% complete for ENCFF000ASM.fastq
Approx 15% complete for ENCFF000ASM.fastq
Approx 20% complete for ENCFF000ASM.fastq
Approx 25% complete for ENCFF000ASM.fastq
Approx 30% complete for ENCFF000ASM.fastq
Approx 35% complete for ENCFF000ASM.fastq
Approx 40% complete for ENCFF000ASM.fastq
Approx 45% complete for ENCFF000ASM.fastq
Approx 50% complete for ENCFF000ASM.fastq
Approx 55% complete for ENCFF000ASM.fastq
Approx 60% complete for ENCFF000ASM.fastq
Approx 65% complete for ENCFF000ASM.fastq
Approx 70% complete for ENCFF000ASM.fastq
Approx 75% complete for ENCFF000ASM.fastq
Approx 80% complete for ENCFF000ASM.fastq
Approx 85% complete for ENCFF000ASM.fastq
Approx 90% complete for ENCFF000ASM.fastq
Approx 95% complete for ENCFF000ASM.fastq
Analysis complete for ENCFF000ASM.fastq
Started analysis of ENCFF000ATK.fastq
Approx 5% complete for ENCFF000ATK.fastq
Approx 10% complete for ENCFF000ATK.fastq
Appr

Далее необходимо проанализировать выдачу fastqc, произвести фильтрацию или подрезание чтений в случае необходимости. Для примера указано использование trimmomatic, также можно использовать seqtk.

In [20]:
!trimmomatic SE -phred33 ENCFF000ARK.fastq ENCFF000ARK_trimmed.fastq ILLUMINACLIP:TruSeq3-SE:2:30:10 \
  LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36

TrimmomaticSE: Started with arguments:
 -phred33 ENCFF000ARK.fastq ENCFF000ARK_trimmed.fastq ILLUMINACLIP:TruSeq3-SE:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36
Automatically using 2 threads
java.io.FileNotFoundException: /content/TruSeq3-SE (No such file or directory)
	at java.io.FileInputStream.open0(Native Method)
	at java.io.FileInputStream.open(FileInputStream.java:195)
	at java.io.FileInputStream.<init>(FileInputStream.java:138)
	at org.usadellab.trimmomatic.fasta.FastaParser.parse(FastaParser.java:54)
	at org.usadellab.trimmomatic.trim.IlluminaClippingTrimmer.loadSequences(IlluminaClippingTrimmer.java:110)
	at org.usadellab.trimmomatic.trim.IlluminaClippingTrimmer.makeIlluminaClippingTrimmer(IlluminaClippingTrimmer.java:71)
	at org.usadellab.trimmomatic.trim.TrimmerFactory.makeTrimmer(TrimmerFactory.java:32)
	at org.usadellab.trimmomatic.Trimmomatic.createTrimmers(Trimmomatic.java:59)
	at org.usadellab.trimmomatic.TrimmomaticSE.run(TrimmomaticSE.java:318)
	at org.u

In [21]:
!./FastQC/fastqc ENCFF000ARK_trimmed.fastq

Started analysis of ENCFF000ARK_trimmed.fastq
Approx 5% complete for ENCFF000ARK_trimmed.fastq
Approx 10% complete for ENCFF000ARK_trimmed.fastq
Approx 15% complete for ENCFF000ARK_trimmed.fastq
Approx 20% complete for ENCFF000ARK_trimmed.fastq
Approx 25% complete for ENCFF000ARK_trimmed.fastq
Approx 30% complete for ENCFF000ARK_trimmed.fastq
Approx 35% complete for ENCFF000ARK_trimmed.fastq
Approx 40% complete for ENCFF000ARK_trimmed.fastq
Approx 45% complete for ENCFF000ARK_trimmed.fastq
Approx 50% complete for ENCFF000ARK_trimmed.fastq
Approx 55% complete for ENCFF000ARK_trimmed.fastq
Approx 60% complete for ENCFF000ARK_trimmed.fastq
Approx 65% complete for ENCFF000ARK_trimmed.fastq
Approx 70% complete for ENCFF000ARK_trimmed.fastq
Approx 75% complete for ENCFF000ARK_trimmed.fastq
Approx 80% complete for ENCFF000ARK_trimmed.fastq
Approx 85% complete for ENCFF000ARK_trimmed.fastq
Approx 90% complete for ENCFF000ARK_trimmed.fastq
Approx 95% complete for ENCFF000ARK_trimmed.fastq
Analy

In [ ]:
!trimmomatic SE -phred33 ENCFF000ATK.fastq ENCFF000ATK_trimmed.fastq ILLUMINACLIP:TruSeq3-SE:2:30:10 \
  LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36

TrimmomaticSE: Started with arguments:
 -phred33 ENCFF000ATK.fastq ENCFF000ATK_trimmed.fastq ILLUMINACLIP:TruSeq3-SE:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36
Automatically using 2 threads
java.io.FileNotFoundException: /content/TruSeq3-SE (No such file or directory)
	at java.io.FileInputStream.open0(Native Method)
	at java.io.FileInputStream.open(FileInputStream.java:195)
	at java.io.FileInputStream.<init>(FileInputStream.java:138)
	at org.usadellab.trimmomatic.fasta.FastaParser.parse(FastaParser.java:54)
	at org.usadellab.trimmomatic.trim.IlluminaClippingTrimmer.loadSequences(IlluminaClippingTrimmer.java:110)
	at org.usadellab.trimmomatic.trim.IlluminaClippingTrimmer.makeIlluminaClippingTrimmer(IlluminaClippingTrimmer.java:71)
	at org.usadellab.trimmomatic.trim.TrimmerFactory.makeTrimmer(TrimmerFactory.java:32)
	at org.usadellab.trimmomatic.Trimmomatic.createTrimmers(Trimmomatic.java:59)
	at org.usadellab.trimmomatic.TrimmomaticSE.run(TrimmomaticSE.java:318)
	at org.u

In [ ]:
!./FastQC/fastqc ENCFF000ATK_trimmed.fastq

Started analysis of ENCFF000ATK_trimmed.fastq
Approx 5% complete for ENCFF000ATK_trimmed.fastq
Approx 10% complete for ENCFF000ATK_trimmed.fastq
Approx 15% complete for ENCFF000ATK_trimmed.fastq
Approx 20% complete for ENCFF000ATK_trimmed.fastq
Approx 25% complete for ENCFF000ATK_trimmed.fastq
Approx 30% complete for ENCFF000ATK_trimmed.fastq
Approx 35% complete for ENCFF000ATK_trimmed.fastq
Approx 40% complete for ENCFF000ATK_trimmed.fastq
Approx 45% complete for ENCFF000ATK_trimmed.fastq
Approx 50% complete for ENCFF000ATK_trimmed.fastq
Approx 55% complete for ENCFF000ATK_trimmed.fastq
Approx 60% complete for ENCFF000ATK_trimmed.fastq
Approx 65% complete for ENCFF000ATK_trimmed.fastq
Approx 70% complete for ENCFF000ATK_trimmed.fastq
Approx 75% complete for ENCFF000ATK_trimmed.fastq
Approx 80% complete for ENCFF000ATK_trimmed.fastq
Approx 85% complete for ENCFF000ATK_trimmed.fastq
Approx 90% complete for ENCFF000ATK_trimmed.fastq
Approx 95% complete for ENCFF000ATK_trimmed.fastq
Analy

In [ ]:
!trimmomatic SE -phred33 ENCFF000ASM.fastq ENCFF000ASM_trimmed.fastq ILLUMINACLIP:TruSeq3-SE:2:30:10 \
  LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36

TrimmomaticSE: Started with arguments:
 -phred33 ENCFF000ASM.fastq ENCFF000ASM_trimmed.fastq ILLUMINACLIP:TruSeq3-SE:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 MINLEN:36
Automatically using 2 threads
java.io.FileNotFoundException: /content/TruSeq3-SE (No such file or directory)
	at java.io.FileInputStream.open0(Native Method)
	at java.io.FileInputStream.open(FileInputStream.java:195)
	at java.io.FileInputStream.<init>(FileInputStream.java:138)
	at org.usadellab.trimmomatic.fasta.FastaParser.parse(FastaParser.java:54)
	at org.usadellab.trimmomatic.trim.IlluminaClippingTrimmer.loadSequences(IlluminaClippingTrimmer.java:110)
	at org.usadellab.trimmomatic.trim.IlluminaClippingTrimmer.makeIlluminaClippingTrimmer(IlluminaClippingTrimmer.java:71)
	at org.usadellab.trimmomatic.trim.TrimmerFactory.makeTrimmer(TrimmerFactory.java:32)
	at org.usadellab.trimmomatic.Trimmomatic.createTrimmers(Trimmomatic.java:59)
	at org.usadellab.trimmomatic.TrimmomaticSE.run(TrimmomaticSE.java:318)
	at org.u

In [ ]:
!./FastQC/fastqc ENCFF000ASM_trimmed.fastq

Started analysis of ENCFF000ASM_trimmed.fastq
Approx 5% complete for ENCFF000ASM_trimmed.fastq
Approx 10% complete for ENCFF000ASM_trimmed.fastq
Approx 15% complete for ENCFF000ASM_trimmed.fastq
Approx 20% complete for ENCFF000ASM_trimmed.fastq
Approx 25% complete for ENCFF000ASM_trimmed.fastq
Approx 30% complete for ENCFF000ASM_trimmed.fastq
Approx 35% complete for ENCFF000ASM_trimmed.fastq
Approx 40% complete for ENCFF000ASM_trimmed.fastq
Approx 45% complete for ENCFF000ASM_trimmed.fastq
Approx 50% complete for ENCFF000ASM_trimmed.fastq
Approx 55% complete for ENCFF000ASM_trimmed.fastq
Approx 60% complete for ENCFF000ASM_trimmed.fastq
Approx 65% complete for ENCFF000ASM_trimmed.fastq
Approx 70% complete for ENCFF000ASM_trimmed.fastq
Approx 75% complete for ENCFF000ASM_trimmed.fastq
Approx 80% complete for ENCFF000ASM_trimmed.fastq
Approx 85% complete for ENCFF000ASM_trimmed.fastq
Approx 90% complete for ENCFF000ASM_trimmed.fastq
Approx 95% complete for ENCFF000ASM_trimmed.fastq
Analy

## Выравнивание на хромосому

Следует выбрать одну хромосому, потому что ресурсы google colab ограничены. В случае проблем, таких как Segmentation fault, error 139, следует выбрать хромосому меньшего размера.

Последовательность нуклеотидов хромосомы можно скачать по адресу https://hgdownload.soe.ucsc.edu/goldenPath/hg38/chromosomes/

In [22]:
!wget https://hgdownload.soe.ucsc.edu/goldenPath/hg38/chromosomes/chr14.fa.gz
!gzip -d chr14.fa.gz

--2022-03-10 06:51:27--  https://hgdownload.soe.ucsc.edu/goldenPath/hg38/chromosomes/chr14.fa.gz
Resolving hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)... 128.114.119.163
Connecting to hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)|128.114.119.163|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 29295890 (28M) [application/x-gzip]
Saving to: ‘chr14.fa.gz’

chr14.fa.gz         100%[===================>]  27.94M  17.4MB/s    in 1.6s    

2022-03-10 06:51:29 (17.4 MB/s) - ‘chr14.fa.gz’ saved [29295890/29295890]



In [23]:
%%time
!bowtie2-build chr14.fa chromosome_index

Settings:
  Output files: "chromosome_index.*.bt2"
  Line rate: 6 (line is 64 bytes)
  Lines per side: 1 (side is 64 bytes)
  Offset rate: 4 (one in 16)
  FTable chars: 10
  Strings: unpacked
  Max bucket size: default
  Max bucket size, sqrt multiplier: default
  Max bucket size, len divisor: 4
  Difference-cover sample period: 1024
  Endianness: little
  Actual local endianness: little
  Sanity checking: disabled
  Assertions: disabled
  Random seed: 0
  Sizeofs: void*:8, int:4, long:8, size_t:8
Input files DNA, FASTA:
  chr14.fa
Building a SMALL index
Reading reference sizes
  Time reading reference sizes: 00:00:01
Calculating joined length
Writing header
Reserving space for joined string
Joining reference sequences
  Time to join reference sequences: 00:00:01
bmax according to bmaxDivN setting: 22642037
Using parameters --bmax 16981528 --dcv 1024
  Doing ahead-of-time memory usage test
  Passed!  Constructing with these parameters: --bmax 16981528 --dcv 1024
Constructing suffix-arr

In [24]:
!mkdir bowtie2_res

In [ ]:
%%time
!bowtie2  -p 2 \
          -x chromosome_index \
          -U ENCFF000ASM_trimmed.fastq \
          -S bowtie2_res/res_ASM.sam

40418814 reads; of these:
  40418814 (100.00%) were unpaired; of these:
    33383834 (82.59%) aligned 0 times
    1752589 (4.34%) aligned exactly 1 time
    5282391 (13.07%) aligned >1 times
17.41% overall alignment rate
CPU times: user 9.56 s, sys: 997 ms, total: 10.6 s
Wall time: 18min 21s


In [ ]:
%%time
!bowtie2  -p 2 \
          -x chromosome_index \
          -U ENCFF000ATK_trimmed.fastq \
          -S bowtie2_res/res_ATK.sam

8383067 reads; of these:
  8383067 (100.00%) were unpaired; of these:
    6879801 (82.07%) aligned 0 times
    354648 (4.23%) aligned exactly 1 time
    1148618 (13.70%) aligned >1 times
17.93% overall alignment rate
CPU times: user 3.36 s, sys: 389 ms, total: 3.75 s
Wall time: 6min 26s


In [25]:
%%time
!bowtie2  -p 2 \
          -x chromosome_index \
          -U ENCFF000ARK_trimmed.fastq \
          -S bowtie2_res/res_ARK.sam

6446684 reads; of these:
  6446684 (100.00%) were unpaired; of these:
    5221078 (80.99%) aligned 0 times
    283250 (4.39%) aligned exactly 1 time
    942356 (14.62%) aligned >1 times
19.01% overall alignment rate
CPU times: user 2.14 s, sys: 271 ms, total: 2.41 s
Wall time: 4min 31s


Проанализируйте выдачу bowtie. Почему процент выравниваний получился именно таким?

Имеет смысл для дальнейшего анализа отобрать уникально картированные риды.

In [26]:
import pandas as pd

df = pd.DataFrame({
        'ChIP-seq': ['ENCFF000ASM_trimmed', 'ENCFF000ATK_trimmed','ENCFF000ARK_trimmed'],
        'Количество ридов всего': [40418814, 8383067, 6446684],
        'Количество ридов, которые выравнились уникально': ["1752589 (4.34%)", "354648 (4.23%)", "283250 (4.39%)"],
        'Количество ридов, которые выравнились неуникально': ["5282391 (13.07%)", "1148618 (13.70%)", "942356 (14.62%)"],
        'Количество ридов, которые не выравнились': ["33383834 (82.59%)", "6879801 (82.07%)", "5221078 (80.99%)"],
     })
df

ChIP-seq  Количество ридов всего  \
0  ENCFF000ASM_trimmed                40418814   
1  ENCFF000ATK_trimmed                 8383067   
2  ENCFF000ARK_trimmed                 6446684   

  Количество ридов, которые выравнились уникально  \
0                                 1752589 (4.34%)   
1                                  354648 (4.23%)   
2                                  283250 (4.39%)   

  Количество ридов, которые выравнились неуникально  \
0                                  5282391 (13.07%)   
1                                  1148618 (13.70%)   
2                                   942356 (14.62%)   

  Количество ридов, которые не выравнились  
0                        33383834 (82.59%)  
1                         6879801 (82.07%)  
2                         5221078 (80.99%)

Преобразование .sam в .bam (не требуется если запускать macs2 на .sam файлах)

In [ ]:
!samtools view -h -S -b -o bowtie2_res/res_ASM.bam bowtie2_res/res_ASM.sam
!samtools view -h -S -b -o bowtie2_res/res_ATK.bam bowtie2_res/res_ATK.sam

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


На этом моменте можно скачать себе на компьютер все полученные .bam или .sam файлы на случай сбоя.

## Peak calling

Помимо macs2 может быть использован пакет Homer для поиска пиков, а также macs3.

In [ ]:
!mkdir macs2

In [ ]:
!macs2 callpeak --broad -t bowtie2_res/res_ASM.sam \
    -c bowtie2_res/res_ATK.sam \
 	  -f SAM \
	  --outdir macs2

INFO  @ Tue, 01 Mar 2022 21:52:11: 
# Command line: callpeak --broad -t bowtie2_res/res_ASM.sam -c bowtie2_res/res_ATK.sam -f SAM --outdir macs2
# ARGUMENTS LIST:
# name = NA
# format = SAM
# ChIP-seq file = ['bowtie2_res/res_ASM.sam']
# control file = ['bowtie2_res/res_ATK.sam']
# effective genome size = 2.70e+09
# band width = 300
# model fold = [5, 50]
# qvalue cutoff for narrow/strong regions = 5.00e-02
# qvalue cutoff for broad/weak regions = 1.00e-01
# The maximum gap between significant sites is assigned as the read length/tag size.
# The minimum length of peaks is assigned as the predicted fragment length "d".
# Larger dataset will be scaled towards smaller dataset.
# Range for calculating regional lambda is: 1000 bps and 10000 bps
# Broad region calling is on
# Paired-End mode is off
 
INFO  @ Tue, 01 Mar 2022 21:52:11: #1 read tag files... 
INFO  @ Tue, 01 Mar 2022 21:52:11: #1 read treatment tags... 
INFO  @ Tue, 01 Mar 2022 21:52:15:  3000000 
INFO  @ Tue, 01 Mar 2022 21:52

In [ ]:
!wget https://www.encodeproject.org/files/ENCFF453PEP/@@download/ENCFF453PEP.bed.gz
!gzip -d ENCFF453PEP.bed.gz

--2022-03-01 22:04:48--  https://www.encodeproject.org/files/ENCFF453PEP/@@download/ENCFF453PEP.bed.gz
Resolving www.encodeproject.org (www.encodeproject.org)... 34.211.244.144
Connecting to www.encodeproject.org (www.encodeproject.org)|34.211.244.144|:443... connected.
HTTP request sent, awaiting response... 307 Temporary Redirect
Location: https://encode-public.s3.amazonaws.com/2017/01/11/2d0a7961-8edc-47f0-a32a-d9e1f4e5455f/ENCFF453PEP.bed.gz?response-content-disposition=attachment%3B%20filename%3DENCFF453PEP.bed.gz&AWSAccessKeyId=ASIATGZNGCNX6DYX23QR&Signature=4lEK8957V7wsopG8xgQvaVXQMLQ%3D&x-amz-security-token=IQoJb3JpZ2luX2VjELb%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLXdlc3QtMiJHMEUCIQC4EYvFRZuHX7WBJFpa6gv%2BduujYCtsThM5WiQcLkDyAAIgIn2xwVlAtmOzjaWmT7PXbyBGH6EMkMEMdo5wcVFI6csq%2BgMIHxAAGgwyMjA3NDg3MTQ4NjMiDFXBKzicxAhN%2FyhbiSrXA6V%2FSjOAmFZ0nwqKpOiOB28o94kybwRlj8U89boOXOfyZwHQ586aGQnX1gTzes4BZaBfcQL6s6Eqhr3uejG5IqZWSe33gj5Bxi0NjcwJglkVn%2FN7cKWiTZuk8AJ2krTkmrXWXXSLFLaRSfgg2STBvAhhmZA

## Сравнение результатов

Сравниваем те пики, которые мы получили, с пиками, которые приведены в ENCODE (важно, чтобы версии генома hg38 или hg19) для .bed файла из ENCODE и той хромосомы, которую скачивали выше, сопадали.

Проанализируйте полученные результаты и приведите свои рассуждения в README.md. Как можно объяснить различия в количестве пересечений?

In [ ]:
!intervene venn -i macs2/NA_peaks.broadPeak ENCFF453PEP.bed --filenames --output venn_results/venn1.jpg


Generating a 2-way "venn" diagram. Please wait...


Done! Please check your results @ venn_results/venn1.jpg. 
Thank you for using Intervene!



In [ ]:
!intervene venn -i ENCFF453PEP.bed macs2/NA_peaks.broadPeak --filenames --output venn_results/venn2.jpg


Generating a 2-way "venn" diagram. Please wait...


Done! Please check your results @ venn_results/venn2.jpg. 
Thank you for using Intervene!

